src/functions.date.js

```diff
@@ -108,28 +108,56 @@ for (let unit in s) {
 	}, {multiValued: true});
 }
 
-_.duration = $.extend(function($this, ms) {
+_.duration = $.extend(function ($this, ms, terms) {
 	if (arguments.length === 1) {
 		[ms, $this] = [$this, null];
 	}
 
-	var count = ms || 0;
-	var unit = "ms";
+	if (ms===0 || terms === undefined) {
+		terms = 1;
+	}
+
+	let timeLeft = ms || 0;
+	let ret = [];
+	let unitsToUse = ["ms"];
 
+	// populate unitsToUse with all the terms that can fit within timeLeft
+	// in order of lowest to highest magnitude,
+	// so we have a list of consecutive unit terms to later reference
 	for (let nextUnit in s) {
-		var nextCount = _.msTo(nextUnit, ms);
+		let count = _.msTo(nextUnit, timeLeft);
 
-		if (nextCount === 0) {
+		if (count === 0) {
 			break;
 		}
 
-		count = nextCount;
-		unit = nextUnit;
+		unitsToUse.push(nextUnit);
 	}
 
-	unit = count === 1 && unit !== "ms"? unit.slice(0, -1) : unit;
+	while (ret.length < terms && unitsToUse.length > 0) {
+		let unit = unitsToUse.pop();
+		let unitTime;
+
+		if (unit!=="ms") {
+			unitTime = _.msTo(unit, timeLeft);
+		}
+		else {
+			unitTime = timeLeft;
+		}
+
+		if (unitTime!=0 || terms === 1) {
+			let unitProperPlurality = unitTime === 1 && unit !== "ms" ? unit.slice(0, -1) : unit;
+			ret.push(unitTime + " " + _.phrase($this, unitProperPlurality));
+			timeLeft -= unitTime * (unit === "ms" ? 1 : Mavo.Functions[unit]());
+		}
+		else {
+			// we will not populate the returning array further since
+			// the following units won't be consecutive
+			break;
+		}
+	}
 
-	return count + " " + _.phrase($this, unit);
+	return (arguments.length ===1 || (arguments.length ===2 && $this!== undefined))? ret[0] : ret;
```

Comment:
Perhaps a variable for the condition here to make it more clear what we're checking?